### PR TITLE
Feature/multigroup part2

### PR DESCRIPTION
**Pull request by whaeck**
This is basically an example of what has to be changed to add a single record.

A side note: eventually, we should consolidate records in other classes by "logic" (e.g. all nubar data in a nubar class, all fission data in a fission class, all group structures in a group structure class, etc.). This'll make creating files so much easier.

### CODE REVIEW

File: src/NDItk/MultigroupTable/src/ctor.hpp [hunk 1]
```diff
@@ -18,19 +19,22 @@ MultigroupTable() : metadata_(), structure_(), weights_(), xs_() {}
...
+                 multigroup::ReactionCrossSections xs,
+                 std::optional< multigroup::AverageFissionEnergyRelease > release = std::nullopt ) :
     metadata_( std::move( zaid ), std::move( libname ), std::move( source ),
                std::move( process ), awr, weight, temperature, dilution,
                xs.numberGroups(), xs.numberReactions() ),
     structure_( std::move( structure ) ),
     weights_( std::move( weigths ) ),
-    xs_( std::move( xs ) ) {
+    xs_( std::move( xs ) ),
+    release_( std::move( release.value() ) ) {
```
Comment (tgsaller) [inline src/NDItk/MultigroupTable/src/ctor.hpp:37]: Why is this .value()? Because it's optional?

File: src/NDItk/MultigroupTable/test/MultigroupTable.test.cpp [hunk 2]
```diff
@@ -35,10 +35,13 @@ SCENARIO( "MultigroupTable" ) {
       multigroup::FluxWeights weights( { 0.1, 0.2, 0.25, 0.05, 0.15, 0.04, 0.06 } );
       multigroup::ReactionCrossSections xs( { { 2, 0.0, { 10., 20., 30., 40., 50., 60., 70. } },
                                               { 16, 1.1234567, { 1., 2., 3., 4., 5., 6., 7. } } } );
+      multigroup::AverageFissionEnergyRelease release( 202.827, 181.238898, 4.827645,
```
Comment (tgsaller) [inline src/NDItk/MultigroupTable/test/MultigroupTable.test.cpp:38]: Is it okay that this stuff is in a different order than in L150?

File: src/NDItk/multigroup/AverageFissionEnergyRelease.hpp [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+
+  /* auxiliary functions */
+
+public:
+
+  /* constructor */
+
+  #include "NDItk/multigroup/AverageFissionEnergyRelease/src/ctor.hpp"
+
+  /* methods */
```
Comment (tgsaller) [inline src/NDItk/multigroup/AverageFissionEnergyRelease.hpp:30]: Can we also have one to return all six as a vector?

File: src/NDItk/multigroup/AverageFissionEnergyRelease/test/AverageFissionEnergyRelease.test.cpp [hunk 4]
```diff
@@ -0,0 +1,106 @@
...
+
+// other includes
+
+// convenience typedefs
+using namespace njoy::NDItk;
+using AverageFissionEnergyRelease = multigroup::AverageFissionEnergyRelease;
+
+std::string chunk();
+void verifyChunk( const AverageFissionEnergyRelease& );
+std::string chunkWithInsufficientNumberBoundaries();
```
Comment (tgsaller) [inline src/NDItk/multigroup/AverageFissionEnergyRelease/test/AverageFissionEnergyRelease.test.cpp:17]: These are from the group boundaries object.

**Review by tgsaller — COMMENTED** [review]
Sorry if some of these things are addressed in later MRs.

File: src/NDItk/MultigroupTable/src/ctor.hpp (hunk 1 above)
Comment (whaeck) [reply src/NDItk/MultigroupTable/src/ctor.hpp:37]: Yes. You use value() to extract the underlying value - if the optional is not a nullopt. In that case, you'll get an exception.

Comment (whaeck) [reply src/NDItk/MultigroupTable/src/ctor.hpp:37]: Which in this case could happen so this is badly coded. I changed this in fix/review-part2

File: src/NDItk/MultigroupTable/test/MultigroupTable.test.cpp (hunk 2 above)
Comment (whaeck) [reply src/NDItk/MultigroupTable/test/MultigroupTable.test.cpp:38]: Yes, since they get rearranged under the hood. The order I used here is:
```
 *  @param[in] total       the total energy release (including delayed particles
 *                         and neutrinos)
 *  @param[in] prompt      the recoverable energy release (total minus delayed
 *                         particles and neutrinos)
 *  @param[in] neutrons    the energy release through prompt neutrons
 *  @param[in] gammas      the energy release through prompt gammas
 *  @param[in] betas       the energy release through delayed betas
 *  @param[in] fragments   the kinetic energy of fission fragments
```

File: src/NDItk/multigroup/AverageFissionEnergyRelease.hpp (hunk 3 above)
Comment (whaeck) [reply src/NDItk/multigroup/AverageFissionEnergyRelease.hpp:30]: Yes, that is the values() method that is taken over from the base class.

File: src/NDItk/multigroup/AverageFissionEnergyRelease/test/AverageFissionEnergyRelease.test.cpp (hunk 4 above)
Comment (whaeck) [reply src/NDItk/multigroup/AverageFissionEnergyRelease/test/AverageFissionEnergyRelease.test.cpp:17]: I removed the unused declarations in fix/review-part3

File: src/NDItk/MultigroupTable/test/MultigroupTable.test.cpp (hunk 2 above)
Comment (tgsaller) [reply src/NDItk/MultigroupTable/test/MultigroupTable.test.cpp:38]: But the order is different if we read from a file versus input it this way?

Comment (whaeck) [reply src/NDItk/MultigroupTable/test/MultigroupTable.test.cpp:38]: The order under the hood is the same as in the file

Comment (whaeck) [reply src/NDItk/MultigroupTable/test/MultigroupTable.test.cpp:38]: Let me rephrase: the data given to the constructor gets rearranged to the order as defined by the NDI format.

File: python/src/MultigroupTable.python.cpp [hunk 5]
```diff
@@ -36,11 +37,13 @@ void wrapMultigroupTable( python::module& module, python::module& ) {
...
                   double, double, double, double,
                   EnergyGroupStructure, FluxWeights,
-                  ReactionCrossSections >(),
+                  ReactionCrossSections,
+                  std::optional< AverageFissionEnergyRelease > >(),
     python::arg( "zaid" ), python::arg( "libname" ), python::arg( "source" ),
     python::arg( "process" ), python::arg( "awr" ), python::arg( "weight" ),
     python::arg( "temperature" ), python::arg( "dilution" ),
     python::arg( "structure" ), python::arg( "flux" ), python::arg( "xs" ),
+    python::arg( "release" ) = std::nullopt,
```
Comment (tgsaller) [inline python/src/MultigroupTable.python.cpp:46]: Could you name this fiss_q to be consistent with the original NDI files?

File: python/src/MultigroupTable.python.cpp [hunk 6]
```diff
@@ -36,11 +37,13 @@ void wrapMultigroupTable( python::module& module, python::module& ) {
     python::init< std::string, std::string, std::string, std::string,
                   double, double, double, double,
                   EnergyGroupStructure, FluxWeights,
-                  ReactionCrossSections >(),
+                  ReactionCrossSections,
+                  std::optional< AverageFissionEnergyRelease > >(),
     python::arg( "zaid" ), python::arg( "libname" ), python::arg( "source" ),
     python::arg( "process" ), python::arg( "awr" ), python::arg( "weight" ),
     python::arg( "temperature" ), python::arg( "dilution" ),
     python::arg( "structure" ), python::arg( "flux" ), python::arg( "xs" ),
```
Comment (tgsaller) [inline python/src/MultigroupTable.python.cpp:45]: I don't like "xs" for the reaction XSs because when the total XS comes in, it might get confusing. Maybe reactionxs?

File: src/NDItk.hpp [hunk 7]
```diff
@@ -3,5 +3,6 @@
 #include "NDItk/multigroup/FluxWeights.hpp"
 #include "NDItk/multigroup/CrossSection.hpp"
 #include "NDItk/multigroup/ReactionCrossSections.hpp"
+#include "NDItk/multigroup/FissionEnergyRelease.hpp"
```
Comment (tgsaller) [inline src/NDItk.hpp:6]: Here it's FissionEnergyRelease while below it's AverageFissionEnergyRelease. Typo?

File: src/NDItk/multigroup/AverageFissionEnergyRelease/src/ctor.hpp [hunk 8]
```diff
@@ -0,0 +1,25 @@
...
+ */
+AverageFissionEnergyRelease() : RealListRecord( "fiss_q" ) {}
+
+
+/**
+ *  @brief Constructor for the primary group structure
+ *
+ *  @param[in] total       the total energy release (including delayed particles
+ *                         and neutrinos)
+ *  @param[in] prompt      the recoverable energy release (total minus delayed
```
Comment (tgsaller) [inline src/NDItk/multigroup/AverageFissionEnergyRelease/src/ctor.hpp:12]: This is labeled "recoverable" below.

File: python/src/MultigroupTable.python.cpp (hunk 6 above)
Comment (whaeck) [reply python/src/MultigroupTable.python.cpp:45]: I changed this following your remark to reaction_xs

File: src/NDItk.hpp (hunk 7 above)
Comment (whaeck) [reply src/NDItk.hpp:6]: Yes, that's the old name so this is wrong. I changed the name of the class to AverageXXXX in case we wanted to add multigroup energy release components one day.

File: src/NDItk/multigroup/AverageFissionEnergyRelease/src/ctor.hpp (hunk 8 above)
Comment (whaeck) [reply src/NDItk/multigroup/AverageFissionEnergyRelease/src/ctor.hpp:12]: prompt probably the better name here so I'll replace recverable

File: python/src/MultigroupTable.python.cpp (hunk 5 above)
Comment (whaeck) [reply python/src/MultigroupTable.python.cpp:46]: I'll change it to fission_q

**Review by tgsaller — APPROVED** [review]
Assuming later MRs address my comments, I think it's good!
